src/features/codeLensProvider.ts

```diff
@@ -6,171 +6,271 @@
-        if (options.showTestsCodeLens) {
-            await this._updateCodeLensForTest(bucket, fileName, node);
-        }
+        return codeLens;
     }
 
-    resolveCodeLens(codeLens: vscode.CodeLens, token: vscode.CancellationToken): Thenable<vscode.CodeLens> {
-        if (codeLens instanceof OmniSharpCodeLens) {
+    private async resolveTestCodeLens(codeLens: TestCodeLens, singularTitle: string, singularCommandName: string, pluralTitle: string, pluralCommandName: string): Promise<vscode.CodeLens> {
+        if (!codeLens.isTestContainer) {
+            // This is just a single test method, not a container.
+            codeLens.command = {
+                title: singularTitle,
+                command: singularCommandName,
+                arguments: [codeLens.testMethodNames[0], codeLens.fileName, codeLens.testFramework]
+            };
 
-            let req = <protocol.FindUsagesRequest>{
-                FileName: codeLens.fileName,
-                Line: codeLens.range.start.line + 1,
-                Column: codeLens.range.start.character + 1,
-                OnlyThisFile: false,
-                ExcludeDefinition: true
+            return codeLens;
+        }
+        
+        const projectInfo = await serverUtils.requestProjectInformation(this._server, { FileName: codeLens.fileName });
+
+        // We do not support running all tests on legacy projects.
+        if (projectInfo.MsBuildProject && !projectInfo.DotNetProject) {
+            codeLens.command = {
+                title: pluralTitle,
+                command: pluralCommandName,
+                arguments: [codeLens.testMethodNames, codeLens.fileName, codeLens.testFramework]
             };
+        }
 
-            return serverUtils.findUsages(this._server, req, token).then(res => {
-                if (!res || !Array.isArray(res.QuickFixes)) {
-                    return;
-                }
+        return codeLens;
+    }
+}
+
+function createCodeLenses(elements: Structure.CodeElement[], fileName: string, options: Options): vscode.CodeLens[] {
+    let results: vscode.CodeLens[] = [];
+
+    Structure.walkCodeElements(elements, element => {
+        let codeLenses = createCodeLensesForElement(element, fileName, options);
+
+        results.push(...codeLenses);
+    });
 
-                let len = res.QuickFixes.length;
-                codeLens.command = {
-                    title: len === 1 ? '1 reference' : `${len} references`,
-                    command: 'editor.action.showReferences',
-                    arguments: [vscode.Uri.file(req.FileName), codeLens.range.start, res.QuickFixes.map(toLocation)]
-                };
+    return results;
+}
+
+function createCodeLensesForElement(element: Structure.CodeElement, fileName: string, options: Options): vscode.CodeLens[] {
+    let results: vscode.CodeLens[] = [];
 
-                return codeLens;
-            });
+    if (options.showReferencesCodeLens && isValidElementForReferencesCodeLens(element)) {
```

Comment:
Is it possible to add a test to verify that we show the codelenses  only when the appropriate option is set to true ?

Reply:
Sure. I'll give that a shot.

Reply:
It looks like this might be a little involved. It looks like the infrastructure is currently designed so that  _all_ integration tests run on both the 'singleCsproj' test asset and the 'slnWithCsproj' test asset. However, the only workspace with a test project is 'slnWithCsproj', so I need to write a test that specifically targets that. Since coverage for this option already didn't exist, are you OK if I merge without it and file an issue?

Reply:
Yup, we can add that later.

Reply:
I went ahead and did it anyway. It's not amazing code, but it works. :smile: